### PR TITLE
HLD for reclaiming the reserved buffer

### PR DESCRIPTION
**Pull request by stephenxs**
Signed-off-by: Stephen Sun <stephens@nvidia.com>

### CODE REVIEW

File: doc/qos/reclaim-reserved-buffer.md [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+
+    attr.id = SAI_QUEUE_ATTR_BUFFER_PROFILE_ID;
+    attr.value.oid = SAI_NULL_OBJECT_ID;
+    sai_queue_api->set_queue_attribute(queue_id, &attr); // queue_id is the SAI object ID of the queue
+
+After this SAI API called, the reserved buffer of the queue indicated by pg_id will be set to zero.
+
+#### Reclaim port reserved buffers ####
+
+The SAI API `sai_port_api->set_port_attribute` is used for reclaiming reserved buffer for queues. The arguments should be the following:
```
Comment (itaibaz) [inline doc/qos/reclaim-reserved-buffer.md]: tpo, it's claiming port pool buffers

Comment (stephenxs) [reply doc/qos/reclaim-reserved-buffer.md]: Thanks. Fixed.

File: doc/qos/reclaim-reserved-buffer.md [hunk 2]
```diff
@@ -0,0 +1,696 @@
...
+   - There is no buffer configuration in SONiC
+   - The reserved buffer size in the ASIC is zero
+
+For some of the buffer objects, the SDK default reserved size is not zero. This is to make sure the system works correctly without any buffer configuration in SONiC.
+Now we have same SONiC configuration in 1 and 2 but different reserved buffer size in the ASIC.
+
+To make it clear and consistent, we need the following solution:
+
+- Introduce a series of buffer profiles whose `size` is zero and `dynamic_th` is -8. This means it will neither reserve buffer nor occupy shared buffer pool. This kind of profiles will be called as `zero buffer profile` in the following section.
+- The buffer objects will be configured to:
```
Comment (stephenxs) [inline doc/qos/reclaim-reserved-buffer.md]: requirement to SAI

File: doc/qos/reclaim-reserved-buffer.md [hunk 3]
```diff
@@ -0,0 +1,696 @@
...
+  - Buffer queue items in `BUFFER_QUEUE` table
+  - Buffer priority group items in `BUFFER_PG` table
+  - Buffer ingress and egress port profile list in `BUFFER_PORT_INGRESS_PROFILE_LIST` and `BUFFER_PORT_EGRESS_PROFILE_LIST` respectively.
+
+#### 7.1.1 Buffer template update for zero buffer profiles ####
+
+Zero buffer profiles should be defined for ingress/egress and lossless/lossy traffic in the buffer template.
+
+```json
+{%- macro generate_zero_buffer_profiles() %}
```
Comment (stephenxs) [inline doc/qos/reclaim-reserved-buffer.md]: better to deploy only if they are used.

File: doc/qos/reclaim-reserved-buffer.md [hunk 4]
```diff
@@ -0,0 +1,696 @@
...
+2. Currently, there is no buffer object configured on admin-down ports. The zero profiles should be configured explicitly on admin-down ports.
+3. `ingress_lossy_profile` is for the ingress lossy priority group.
+
+   Currently, it contains all zero values but the size in the ASIC will be set to pipeline-latency by SAI/SDK. Now that zero buffer profiles have been introduced, SAI will not be able to distinguish `ingress_lossy_profile` from zero buffer profiles given that both buffer profile have `0` as their `size`.
+
+   There are two solutions to resolve it:
+   - To redefine the `ingress_lossy_profile` with pipeline latency as `size`.
+
+     This is complicated because 8-lane ports have different pipeline latency from other ports. The buffer template can not distinguish them. A standalone script is required to deploy the profile.
+   - SAI to distinguish `ingress_lossy_profile` and zero buffer profiles by using `dynamic_th`. A profile will be treated as zero profile if its `dynamic_th` is `-8`.
```
Comment (stephenxs) [inline doc/qos/reclaim-reserved-buffer.md]: add SAI extension for this.

Comment (stephenxs) [reply doc/qos/reclaim-reserved-buffer.md]: No buffer profile. other fields can be empty.

Comment (stephenxs) [reply doc/qos/reclaim-reserved-buffer.md]: also add attribute in CONFIG_DB

File: doc/qos/reclaim-reserved-buffer.md [hunk 5]
```diff
@@ -0,0 +1,696 @@
...
+  },
+  "BUFFER_PORT_EGRESS_PROFILE_LIST" : {
+    "Ethernet0" : {
+      "profile_list" : "[BUFFER_PROFILE|egress_lossless_zero_profile],[BUFFER_PROFILE|egress_lossy_zero_profile]"
+    }
+  }
+}
+```
+
+#### 7.1.2 The flow to deploy a switch ####
```
Comment (stephenxs) [inline doc/qos/reclaim-reserved-buffer.md:508]: this is not claim. this is to not configure

File: doc/qos/reclaim-reserved-buffer.md (hunk 4 above)
Comment (stephenxs) [reply doc/qos/reclaim-reserved-buffer.md]: No need to add an attribute in `CONFIG_DB`. Just a buffer profile with {`NULL`: `NULL`} only can indicate it's an empty profile.

File: doc/qos/reclaim-reserved-buffer.md [hunk 6]
```diff
@@ -0,0 +1,696 @@
...
+- The buffer objects will be configured to:
+  - SDK default value, if there is no buffer profile configured for the object in SONiC
+  - SDK default value, if the buffer object is removed from SONiC
+  - Zero, only if the object is configured with a zero profile
+
+To achieve it the following steps will be taken:
+
+1. Zero buffer profiles should be defined for ingress/egress and lossless/lossy traffic.
+2. Currently, there is no buffer object configured on admin-down ports. The zero profiles should be configured explicitly on admin-down ports.
+3. `ingress_lossy_profile` is for the ingress lossy priority group.
```
Comment (stephenxs) [inline doc/qos/reclaim-reserved-buffer.md]: Need to find a way to indicate whether a platform supports `empty profile`. Candidates include
- if the `empty_profile` is provided by the vendor, it is supported
  - pros: easy to implement. no timing hazard
  - cons: the `empty_profile` needs to be loaded even if it is not used.
- query SAI capability.
  - cons: 
    - SAI attribute can only be fetched by orchagent and then exposed to state db. 
    - no guarantee that the capability is available when buffer manager starts. need to handle both conditions.

File: doc/qos/reclaim-reserved-buffer.md (hunk 4 above)
Comment (stephenxs) [reply doc/qos/reclaim-reserved-buffer.md]: what will happen if SAI doesn't support it?

**Pull request closed without merging**